# TD 4: Listes

---

## 1. Débuts avec les listes

Dans cette partie, il faut écrire les réponses directement dans le notebook. En plus du cours, n'hésitez pas à consulter la documentation (tutoriel) Python sur les listes [ici (introduction aux listes)](https://docs.python.org/fr/3.6/tutorial/introduction.html#lists) et [ici (complément sur les listes)] (https://docs.python.org/fr/3/tutorial/datastructures.html#more-on-lists).

1. Créer la liste `3, 5, 10`, l'affecter à une variable, puis l'afficher.

In [30]:
l = [3,5,10]
print(l)


[3, 5, 10]


2. Ajouter ensuite 12 et 17 à la fin puis afficher à nouveau la liste.

In [31]:
l.extend([12, 17])
print(l)

[3, 5, 10, 12, 17]


3. Changer le 10 en -7.

In [35]:
l[2] = -7
print(l)

[3, 5, -7, 12, 17]


4. Avec une boucle `for`, écrire le double de chaque élément de la liste. Cela doit afficher `[6, 10, -14, 24, 34]`. La liste doit être modifiée.

In [5]:
for i in range(len(l)):
    l[i] *=2
print(l)


[6, 10, -14, 24, 34]


5. A l'aide d'une boucle `for`, ajouter  la valeur `i` à l'élément d'indice `i` de la liste. Cela doit afficher `[6, 11, -12, 27, 38]`. La liste doit être modifiée.

In [7]:
for i in range(len(l)):
    l[i] += i
print(l)

[6, 11, -12, 27, 38]


6. Tirer 10 nombres à 2 chiffres au hasard et les mettre dans une nouvelle liste (vide au départ), puis afficher la somme des entiers de la liste et le plus grand de ces entiers.

In [52]:

import random
n = [random.randint(10, 99) for i in range(10)]
print(n)
sum = 0 

for i in range (0, len(n)):
    sum += n[i]
    
print(sum)
print(max(n))

#ou alors on fait n = [random.randint(10, 99) for i in range(10)] pour créer la liste

[97, 86, 81, 91, 60, 77, 24, 80, 76, 11]


TypeError: 'int' object is not callable

7. A partir de la liste choisie au hasard, créer une liste qui contient les éléments pairs et une deuxième liste qui contient les éléments impairs.

In [49]:
pair, impair = [], []
for elem in n:
    if elem%2 == 0:
        pair.append(elem)
    else:
        impair.append(elem)
print(pair, impair)



[42, 14, 24, 54] [21, 11, 89, 17, 51, 21]


8. Trier la liste choisie au hasard par ordre croissant en modifiant la liste.

In [51]:
n.sort()
print(n)

[11, 14, 17, 21, 21, 24, 42, 51, 54, 89]


9. Supprimer le plus petit et le plus grand élément de la liste triée.

In [11]:
n.remove(min(n))
n.remove(max(n))
print(n)

[34, 36, 42, 49, 51, 60, 80, 94]


10. Créer la liste des entiers de 1 jusqu'à 52. Ces entiers représentent les cartes d'un paquet de 52 cartes.

In [67]:
paquet = list(range(1,53))

 11. Couper le jeu de carte revient à choisir un indice `i` entre 0 et 51, et à passer toutes les cartes allant des indices 0 à i-1 à la suite des cartes qui vont des indices i à 51. Ecrire le code qui effectue une coupe choisie au hasard.

In [73]:
import random
coupe = random.randint(0, 51)
"""
code faux : 

for i in range(coupe):
    paquet.append(paquet[i])
    del(paquet[i])*
"""


paquet = (paquet + paquet[0:coupe])[coupe:]
paquet



[34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 1,
 3,
 5,
 7,
 9,
 11,
 13,
 15,
 17,
 2,
 6,
 10,
 14,
 18,
 20,
 22,
 24,
 26,
 28,
 4,
 12,
 8,
 16,
 19,
 23,
 27,
 30,
 32,
 21,
 25,
 29,
 31,
 33]

---

## 2. Problème de Syracuse

Cette partie est à traiter dans un script que vous pouvez appeler, par exemple, `syracuse.py`.

En partant d'un entier `n` de départ, on définit une suite d'entiers en obtenant chaque nouveau terme à partir du précédent soit en le divisant par 2 s'il est pair, soit en le multipliant par 3 et en ajoutant 1 s'il est impair. 

1. Sans écrire de programme, calculer les premières valeurs de la suite en chosissant `n = 3`.

2. Ecrire la fonction qui, à partir d'un entier initial, renvoie la liste des valeurs successives jusqu'à ce que la valeur `1` soit atteinte (le dernier élément de la liste est donc toujours 1).

In [1]:
def syracuse(n):
    """ Retourne la liste des valeurs de la suite en partant de n jusqu'à 1 """
    l = [n]
    while n != 1:
        if n %2 == 0:
            n = n//2
        else:
            n = n*3 + 1
        l.append(n)
    return l

print(syracuse(3))

[3, 10, 5, 16, 8, 4, 2, 1]


3. La conjecture de Collatz (ou problème de Syracuse) affirme que, quel que soit l’entier `n` que l’on choisisse au départ, on finit toujours par arriver à 1 (ce résultat est une *conjecture*, c'est-à-dire qu'il n'a pas été démontré, mais qu'il n'existe pas de contre-exemple connu). Écrire une fonction qui, en appelant la fonction précédente, va vérifier si la conjecture est vraie pour `n` de 1 à `n_max`, où `n_max` est un paramètre de la fonction. 

*Remarque*: si tout se passe bien, la fonction doit juste se terminer et renvoyer `True` par exemple. Sinon, c’est qu’on sera entré dans une boucle infinie.

In [2]:
def testeConjecture(n_max):
    """ Teste la conjecture de Collatz pour toutes les valeurs de 1 à n_max """
    for i in range(1, n_max + 1):
        return 1 in syracuse(i)

print(testeConjecture(10000))

True


4. On appelle *temps de vol* de l’entier `n` le nombre d’étapes pour aller de `n` jusqu’à 1. Le temps de vol de 1 est 0, le temps de vol de 3 est 7. Écrire une fonction qui, étant donné un paramètre `n`, renvoie son temps de vol.

In [3]:
def tempsVol(n):
    """ Retourne le temps de vol de n """
    return len(syracuse(n)) - 1

print("Le temps de vol de", 3, "est", tempsVol(3))

Le temps de vol de 3 est 7


5. Ecrire une fonction qui, étant donné un paramètre `n_max` renvoie la liste des temps de vol de tous les entiers de 1 à `n_max`. *Indication*: utiliser une liste en compréhension.

In [5]:
def tempsVolListe(n_max):
    """ Retourne la liste de tous les temps de vol de 1 à n_max """
    return [tempsVol(i) for i in range(1, n_max + 1)]

print(tempsVolListe(100))

#question 6
L = tempsVolListe(10000)
t_max = max(L)

print("l'entier", L.index(t_max) + 1, "a le plus grand temps de vol égal à", t_max)

#question 7
def alt_max(n):
    return max(syracuse(n))
def alt_max_liste(alt_max):
    return [alt_max[i] for i in range(1, alt_max + 1)]

L_alt = alt_max_liste(10000)
b = max(L_alt)
print("l'entier", L_alt.index(b) + 1, "a une altitude max de", b)

[0, 1, 7, 2, 5, 8, 16, 3, 19, 6, 14, 9, 9, 17, 17, 4, 12, 20, 20, 7, 7, 15, 15, 10, 23, 10, 111, 18, 18, 18, 106, 5, 26, 13, 13, 21, 21, 21, 34, 8, 109, 8, 29, 16, 16, 16, 104, 11, 24, 24, 24, 11, 11, 112, 112, 19, 32, 19, 32, 19, 19, 107, 107, 6, 27, 27, 27, 14, 14, 14, 102, 22, 115, 22, 14, 22, 22, 35, 35, 9, 22, 110, 110, 9, 9, 30, 30, 17, 30, 17, 92, 17, 17, 105, 105, 12, 118, 25, 25, 25]
l'entier 6171 a le plus grand temps de vol égal à 261


TypeError: 'int' object is not subscriptable

6. Déterminer quel entier entre 1 et 10000 a le plus grand temps de vol (réponse 6171 qui a le temps de vol 261).

7. L’altitude maximale de l'entier `n` est la plus grande valeur par laquelle passe `n` au cours de son vol. Déterminer quel entier entre 1 et 10000 a la plus grande altitude maximale (réponse 27114424, atteint par l'entier 9663). Ne pas hésiter à écrire de nouvelles fonctions pour cela.

---

## 3. Carré magique

*Définition* extraite de [l'article Wikipedia sur le carré magique](https://fr.wikipedia.org/wiki/Carr%C3%A9_magique_(math%C3%A9matiques)): en mathématiques, un carré magique d’ordre $n$ est composé de $n^2$ entiers strictement positifs, écrits sous la forme d’un tableau carré. Ces nombres sont disposés de sorte que leurs sommes sur chaque rangée, sur chaque colonne et sur chaque diagonale principale soient égales. On nomme alors *constante magique* la valeur de ces sommes. 

Un exemple de carré magique:

$$\begin{bmatrix} 4 & 14 & 15 & 1 \\ 9 & 7 & 6 & 12 \\ 5 & 11 & 10 & 8 \\ 16 & 2 & 3 & 13 \end{bmatrix}$$

1. Sans écrire de programme, vérifier qu'il s'agit bien d'un carré magique, et donner sa constante magique.

Les questions qui suivent doivent être traitées dans un script, qui peut s'appeler `carre_magique.py`. Pensez à utiliser [PythonTutor](http://pythontutor.com/) et les outils de programmation pour résoudre les problèmes qui pourraient survenir.

2. Créer une liste à deux dimensions affectée à  la variable `carre_mag` contenant ce carré magique.

3. Créer une deuxième liste affectée à la variable `carre_pas_mag` qui ne soit pas magique à partir du carré magique en changeant le 3 en 7. 

4. Créer une fonction qui affiche la liste comme un carré.

In [24]:
carre_mag = [[4, 14, 15, 1],[9, 7, 6, 12], [5, 11, 10, 8], [16, 2, 3, 13]]
carre_pas_mag = [i for i in carre_mag]
carre_pas_mag[3][2] = 7 
print(carre_mag)
print(carre_pas_mag)

[[4, 14, 15, 1], [9, 7, 6, 12], [5, 11, 10, 8], [16, 2, 7, 13]]
[[4, 14, 15, 1], [9, 7, 6, 12], [5, 11, 10, 8], [16, 2, 7, 13]]


In [31]:
#carre magique 
carre_mag = [[4, 14, 15, 1],[9, 7, 6, 12], [5, 11, 10, 8], [16, 2, 3, 13]]
carre_pas_mag = [lignes[:] for lignes in carre_mag]

carre_pas_mag[3][2] = 7

def afficheCarre(carre):
    """ Affiche la liste à 2 dimensions carre comme un carré"""
    for i in carre:
        print(i)
    print()

afficheCarre(carre_mag)
afficheCarre(carre_pas_mag)


[4, 14, 15, 1]
[9, 7, 6, 12]
[5, 11, 10, 8]
[16, 2, 3, 13]

[4, 14, 15, 1]
[9, 7, 6, 12]
[5, 11, 10, 8]
[16, 2, 7, 13]



5. Ecrire une fonction qui teste si les lignes du carré ont toutes la même somme. Cette fonction renvoie cette somme si c'est le cas, et `-1` sinon.

In [33]:
def testLignesEgales(carre):
    """ Renvoie la somme des éléments d'une ligne de la liste 2D carre si toutes les lignes ont la même somme, et -1 sinon """
    somme = sum(carre[0])
    for lignes in carre:
        if sum(lignes) != somme:
            return -1
    return somme
    

print(testLignesEgales(carre_mag))
print(testLignesEgales(carre_pas_mag))


34
-1


6. Même question mais en testant les colonnes.

In [37]:
def testColonnesEgales(carre):
    """ Renvoie la somme des éléments d'une colonne de la liste 2D carre si toutes les colonnes ont la même somme, et -1 sinon """
    c1 = [lignes[0] for lignes in carre]
    somme = sum(c1)
    for j in range(1, len(carre)):
        colonne = [lignes[j] for lignes in carre]
        if sum(colonne) != somme:
            return -1 
    return somme

print(testColonnesEgales(carre_mag))
print(testColonnesEgales(carre_pas_mag))

34
-1


7. Même question mais en testant les 2 diagonales.

In [38]:
def testDiagonalesEgales(carre):
    """ Renvoie la somme des éléments d'une diagonale de la liste 2D carre si les 2 diagonales ont la même somme, et -1 sinon """
    taille = len(carre)
    diag1 = [carre[i][i] for i in range(taille)]
    diag2 = [carre[i][taille-i-1] for i in range(taille)]
    somme = sum(diag1)
    if sum(diag2) != somme:
        return -1
    else:
        return somme

print(testDiagonalesEgales(carre_mag))
print(testDiagonalesEgales(carre_pas_mag))

34
34


8. Créer une fonction qui teste si un carré est magique. Elle retourne `True` si c'est le cas et `False` sinon.

In [39]:
def estCarreMagique(carre):
    """ Renvoie True si c'est un carre magique et False sinon"""
    return testLignesEgales(carre) == testColonnesEgales(carre) and testLignesEgales(carre) == testDiagonalesEgales(carre) and testLignesEgales != -1

print(estCarreMagique(carre_mag))
print(estCarreMagique(carre_pas_mag))

True
False


9. Un carré d'ordre $n$ est *normal* s'il contient tous les entiers de 1 à $n^2$. Ecrire une fonction qui teste si un carré est normal (pas nécessairement magique).

In [41]:
def estNormal(carre):
    """ Retourne True si contient toutes les valeurs de 1 à n^2 où n est la taille 
        du carré, et False sinon """
    liste = []
    for lignes in carre:
        liste.extend(lignes)
    taille = len(carre)
    for entier in range(1, taille*taille + 1):
        if entier not in liste:
            return False
    return True 

print(estNormal(carre_mag))
print(estNormal(carre_pas_mag))

True
False
